proposals/yield-in-try-catch.md

```diff
@@ -0,0 +1,500 @@
+
+For example this is the `Dispose` method for the above iterator:
+
+```csharp
+[DebuggerHidden]
+void IDisposable.Dispose()
+{
+    int num = <>1__state;
+    if ((uint)(num - -4) > 1u && num != 1)
+    {
+        return;
+    }
+    try
+    {
+        if (num != -4 && num != 1)
+        {
+            return;
+        }
+        try
+        {
+        }
+        finally
+        {
+            <>m__Finally2();
+        }
+    }
+    finally
+    {
+        <>m__Finally1();
+    }
+}
+```
+
+This behavior is important to understand when considering the code generation for `try / catch` blocks.
+
+### Code generation of yield inside try with catch in iterators
+
+The code generation for iterators that have `yield` inside a `try / catch` will preserve the `catch` semantics in the `Dispose` method. That is _only_ the contents of the `finally` blocks will execute even in the case exceptions come into play. For example:
+
+```csharp
+var e = M1(true).GetEnumerator();
+e.MoveNext();
+e.Dispose();
+
+IEnumerable<int> M1(bool b)
+{
+    try
+    {
+        try
+        {
+            try
+            {
+                yield return 13;
+            }
+            finally
+            {
+                Console.WriteLine("inner finally");
+                if (b)
+                {
+                    throw new Exception();
+                }
+            }
+        }
+        catch
+        {
+            Console.WriteLine("catch");
+        }
+
+        Console.WriteLine("after catch");
+    }
+    finally
+    {
+        Console.WriteLine("outer finally");
+    }
+}
+```
+
+This program will output:
+
+```cmd
+inner finally
+outer finally
+Unhandled exception. System.Exception: Exception of type 'System.Exception' was thrown
+```
+
+The `"after catch"` is not printed beacuse only the `finally` structure is mirrored in the `Dispose` method. The `catch`, like all other statements between `finally` is not included. Statements in between the `catch` and `finally` are not executed in `Dispose`. This may seem odd at first glance but is leaning into the specified behavior for iterator `Dispose`.
+
+### Dispose and finally in async iterators
+
+The `DiposeAsync` behavior for async iterators mirrors that of traditional iterators in that it executes the `finally` blocks at the the point the state machine is suspended. However it does this by setting the `state` variable to a value that represents disposing and then calls `MoveNextAsync`. The implementation of `MoveNextAsync` will then execute only the `finally` blocks for the suspend point. The `DisposeAsync` method does not have a mirror copy of the `finally` blocks.
```

Comment:
>it does this by setting the `state` variable [](http://example.com/codeflow?start=183&length=44)

It's not the "state" variable that indicates disposal mode, but a dedicated variable. It's called "disposeMode" in the speclet: [https://github.com/dotnet/roslyn/blob/main/docs/features/async-streams.md#detailed-design-for-async-iterator-methods](https://github.com/dotnet/roslyn/blob/main/docs/features/async-streams.md#detailed-design-for-async-iterator-methods)